### PR TITLE
#4456 [Manager] Migrate topic to redux with rxjs

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4456)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/components/Pipeline/Pipeline.js [hunk 1]
```diff
@@ -264,7 +264,7 @@ const Pipeline = React.forwardRef((props, ref) => {
       clean({ from: streamCell, paperApi });
     });
 
-    await removeTopic(topic, paperApi);
+    await stopAndRemoveTopic(topic, paperApi);
```
Comment (eechih) [inline ohara-manager/client/src/components/Pipeline/Pipeline.js]: await is redundant

File: ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js [hunk 2]
```diff
@@ -19,11 +19,10 @@ import _ from 'lodash';
 
 import * as hooks from 'hooks';
 import { KIND, CELL_STATUS } from 'const';
-import { useTopicState } from 'context';
 import { PaperContext } from '../Pipeline';
 
 export const useTopics = () => {
-  const { data: topicsData } = useTopicState();
+  const topicsData = hooks.useCurrentTopics();
```
Comment (eechih) [inline ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js]: useCurrentTopics 好像有點隱晦 :( 

- useTopicsInWorkspace
- useTopicsInPipeline 

這樣是否好一點？

Comment (saivirtue) [reply ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js]: useTopicsInWorkspace
useTopicsInPipeline
這兩個的result應該是一樣的?

我改成
- useTopicsInWorkspace
- useAllTopics
如何?

Comment (eechih) [reply ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js]: workspace 和 pipeline 的 topics 有一樣嗎？範圍不一樣大吧～

Comment (eechih) [reply ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js]: useTopicsInPipeline 是指被組合在在某個 pipeline 內的 topics，而 toolbox 上用的 topics 應該是 useTopicsInWorkspace

Comment (saivirtue) [reply ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js]: 因為我們topic : group的定義為 `hash(workspaceName, workspaceGroup)`，目前沒有辦法區分topic是在哪一個pipeline內使用 (現在Pipeline內使用的是該workspace下所有的topic)
不過這倒提醒我一件事: pipelineOnly的topic不能被共用。我修改一下這一段邏輯

Comment (eechih) [reply ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js]: 另外，再增加一組 useAllTopics 可以取得全部的 topics（全部 workspace 下的 topics）

Comment (eechih) [reply ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js]: > 因為我們topic : group的定義為 hash(workspaceName, workspaceGroup)，目前沒有辦法區分topic是在哪一個pipeline內使用 (現在Pipeline內使用的是該workspace下所有的topic)

可以由當前的 pipeline 反向查找 topics

Comment (saivirtue) [reply ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js]: Have updated the actions:
- useAllTopics
- useTopicsInWorkspace
- useTopicsInToolbox
- useTopicsInPipeline
- useTopic

Comment (eechih) [reply ohara-manager/client/src/components/Pipeline/Toolbox/ToolboxHooks.js]: 👍

File: ohara-manager/client/src/hooks/topicHooks.js [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+/**
+ * Filter the topics which are displayed in current pipeline
+ *
+ * Note: We will only show topics which are added in paper
+ */
+export const useTopicsInPipeline = () => {
+  const pipelineGroup = hooks.usePipelineGroup();
+  const pipelineName = hooks.usePipelineName();
+  const id = getId({ group: pipelineGroup, name: pipelineName });
+  const getPipelineById = useMemo(selectors.makeGetPipelineById, []);
```
Comment (eechih) [inline ohara-manager/client/src/hooks/topicHooks.js:179]: 應該可以直接使用 hooks.usePipeline()

File: ohara-manager/client/src/hooks/topicHooks.js [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+export const useTopic = name => {
+  const getTopicById = useMemo(selectors.makeGetTopicById, []);
+  const getInfoById = useMemo(selectors.makeGetInfoById, []);
+  const group = useTopicGroup();
+  const id = getId({ group, name });
+  return useSelector(
+    useCallback(
+      state => {
+        const topic = getTopicById(state, { id });
+        const info = getInfoById(state, { id });
```
Comment (eechih) [inline ohara-manager/client/src/hooks/topicHooks.js:215]: 🤔 有點好奇 info 會有資料嗎？直覺 id 好像不對

File: ohara-manager/client/src/hooks/topicHooks.js (hunk 3 above)
Comment (saivirtue) [reply ohara-manager/client/src/hooks/topicHooks.js:179]: you are right :)

File: ohara-manager/client/src/hooks/topicHooks.js (hunk 4 above)
Comment (saivirtue) [reply ohara-manager/client/src/hooks/topicHooks.js:215]: Uh...forgot to refactor this :(

**Comment by saivirtue** [comment]
retry manager-e2e

**Comment by saivirtue** [comment]
The failed tests are related to broken add node card (#4470). It will be fixed in #4049.
I will merge this PR if there were no other concerns.

**Pull request merged**